common/unified/solver/gcr_kernels.cpp

```diff
@@ -0,0 +1,128 @@
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/solver/gcr_kernels.hpp"
+
+
+#include <ginkgo/core/base/math.hpp>
+
+
+#include "common/unified/base/kernel_launch_solver.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace GKO_DEVICE_NAMESPACE {
+/**
+ * @brief The GCR solver namespace.
+ *
+ * @ingroup grc
+ */
+namespace gcr {
+
+
+template <typename ValueType>
+void initialize(std::shared_ptr<const DefaultExecutor> exec,
+                const matrix::Dense<ValueType>* b,
+                matrix::Dense<ValueType>* residual,
+                stopping_status* stop_status)
+{
+    run_kernel_solver(
+        exec,
+        [] GKO_KERNEL(auto row, auto col, auto b, auto residual, auto stop) {
+            if (row == 0) {
+                stop[col].reset();
+            }
+            // TODO: possibly set other vals to zero here?
+            residual(row, col) = b(row, col);
+        },
+        // Note: default_stride only applied to objects created using
+        // create_with_config_of as this guarantees identical stride.
+        b->get_size(), b->get_stride(), b, residual, stop_status);
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_GCR_INITIALIZE_KERNEL);
+
+
+template <typename ValueType>
+void restart(std::shared_ptr<const DefaultExecutor> exec,
+             const matrix::Dense<ValueType>* residual,
+             matrix::Dense<ValueType>* A_residual,
+             matrix::Dense<ValueType>* p_bases,
+             matrix::Dense<ValueType>* Ap_bases, size_type* final_iter_nums)
+{
+    run_kernel_solver(
+        exec,
+        [] GKO_KERNEL(auto row, auto col, auto residual, auto A_residual,
+                      auto p_bases, auto Ap_bases, auto final_iter_nums) {
+            if (row == 0) {
+                final_iter_nums[col] = 0;
+            }
+            p_bases(row, col) = residual(row, col);
+            Ap_bases(row, col) = A_residual(row, col);
+        },
+        residual->get_size(), residual->get_stride(), residual, A_residual,
+        p_bases, Ap_bases, final_iter_nums);
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_GCR_RESTART_KERNEL);
+
+
+template <typename ValueType>
+void step_1(std::shared_ptr<const DefaultExecutor> exec,
+            matrix::Dense<ValueType>* x, matrix::Dense<ValueType>* residual,
+            const matrix::Dense<ValueType>* p,
+            const matrix::Dense<ValueType>* Ap,
+            const matrix::Dense<remove_complex<ValueType>>* Ap_norm,
+            const matrix::Dense<ValueType>* alpha,
+            const stopping_status* stop_status)
+{
+    run_kernel_solver(
+        exec,
+        [] GKO_KERNEL(auto row, auto col, auto x, auto residual, auto p,
+                      auto Ap, auto Ap_norm, auto alpha, auto stop) {
+            if (!stop[col].has_stopped()) {
+                auto tmp = alpha[col] / Ap_norm[col];
```

Comment:
I mentioned it somewhere else, while this is perfectly fine, changing it to something like
`auto alpha = rAp[col] / Ap_norm[col]` would make it easier to understand. IMO

Reply:
I totally agree with you. I was also confused when I read the code. The code was copied from the previous PR.
I will change that since it's not only me that has the trouble. Thanks for the comment!